.github/workflows/test.yml

```diff
@@ -0,0 +1,245 @@
+          sudo docker cp buildkit:/usr/bin/buildctl /usr/local/bin/
+      -
+        name: Cache binaries
+        uses: actions/cache@v2
+        with:
+          path: /tmp/.buildkit-cache/binaries
+          key: ${{ runner.os }}-buildkit-binaries-${{ github.sha }}
+          restore-keys: |
+            ${{ runner.os }}-buildkit-binaries-
+      -
+        name: Build binaries
+        run: |
+          buildctl build --frontend=dockerfile.v0 \
+            --local context=. --local dockerfile=. \
+            --opt target=binaries \
+            --import-cache type=local,src=/tmp/.buildkit-cache/binaries \
+            --export-cache type=local,dest=/tmp/.buildkit-cache/binaries
+      -
+        name: Dump context
+        if: always()
+        uses: crazy-max/ghaction-dump-context@v1
+
+  integration-tests-base:
+    runs-on: ubuntu-latest
+    steps:
+      -
+        name: Checkout
+        uses: actions/checkout@v2
+      -
+        name: Install buildctl
+        run: |
+          mkdir -p /tmp/buildctl
+          docker run --rm --privileged linuxkit/binfmt:$LINUXKIT_BINFMT
+          docker run --name buildkit -d --privileged -p $BUILDKIT_PORT:$BUILDKIT_PORT $REPO_SLUG_ORIGIN --debug --addr $BUILDKIT_HOST --oci-worker-gc=false
+          sudo docker cp buildkit:/usr/bin/buildctl /usr/local/bin/
+      -
+        name: Cache integration-tests
+        uses: actions/cache@v2
+        with:
+          path: /tmp/.buildkit-cache/integration-tests
+          key: ${{ runner.os }}-buildkit-integration-tests-${{ github.sha }}
+          restore-keys: |
+            ${{ runner.os }}-buildkit-integration-tests-
+      -
+        name: Build integration-tests
+        run: |
+          buildctl build --frontend=dockerfile.v0 \
+            --local context=. --local dockerfile=. \
+            --opt target=integration-tests-base \
+            --import-cache type=local,src=/tmp/.buildkit-cache/integration-tests \
+            --export-cache type=local,dest=/tmp/.buildkit-cache/integration-tests
+      -
+        name: Dump context
+        if: always()
+        uses: crazy-max/ghaction-dump-context@v1
+
+  integration-tests:
+    runs-on: ubuntu-latest
+    needs: [ integration-tests-base ]
+    env:
+      IID: buildkit-tests
+      IID_FILE: /tmp/docker-iidfile
+      VOL: /tmp/.buildkit-cache-test
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          -
+            pkg: ./client
+            flags: -v --timeout=20m
+            skip-it: 0
+          -
+            pkg: ./cmd/buildctl
+            flags: -v
+            skip-it: 0
+          -
+            pkg: ./worker/containerd
+            flags: -v
+            skip-it: 0
+          -
+            pkg: ./...
+            flags: -v
+            skip-it: 1
+          -
+            pkg: ./frontend
+            flags: -v
+            skip-it: 0
+          -
+            pkg: ./frontend/dockerfile
+            flags: -v --parallel=6 --timeout=20m

```

Comment:
I know this is not related to the GH migration, but this test seems like the slowest test, I wonder if we could get better parallelization if we break them out by each buildkit worker and let GH parallelize.  Something like:
```
- pkg: ./frontend/dockerfile
  flags: -v --parallel=6 --timeout=20m -run=".*/.*/worker=containerd$"
  skip-it: 0
- pkg: ./frontend/dockerfile
  flags: -v --parallel=6 --timeout=20m -run=".*/.*/worker=containerd-1.3$"
  skip-it: 0
- pkg: ./frontend/dockerfile
  flags: -v --parallel=6 --timeout=20m -run=".*/.*/worker=containerd-snapshotter-stargz$"
  skip-it: 0
- pkg: ./frontend/dockerfile
  flags: -v --parallel=6 --timeout=20m -run=".*/.*/worker=oci$"
  skip-it: 0
- pkg: ./frontend/dockerfile
  flags: -v --parallel=6 --timeout=20m -run=".*/.*/worker=oci-rootless$"
  skip-it: 0
- pkg: ./frontend/dockerfile
  flags: -v --parallel=6 --timeout=20m -run=".*/.*/worker=oci-snapshotter-stargz$"
  skip-it: 0
```
Not sure if there is a more graceful way to add the list of workers to the matrix.  A similar hack could be applied to all the other packages that run via workers.  If GH can spin up parallel containers for each test I suspect it will reduce the test duration by ~6x.

Reply:
It seems like we can add our own matrix options, so was thinking something like this:
```
matrix:
  pkg:
    - ./client
    - ./cmd/buildctl
    - ./worker/containerd
    - ./frontend
    - ./frontend/dockerfile
  worker: 
    - containerd
    - containerd-1.3
    - containerd-snapshotter-stargz
    - oci
    - oci-rootless
    - oci-snapshotter-stargz
  include:
    - pkg: ./client
      flags: --timeout=20m
    - pkg: ./frontend/dockerfile
      flags: --timeout=20m
steps:
  -  name: Test ${{matrix.worker}}:${{ matrix.pkg }}
     run: |
          ...
          cid=$(docker create --rm -v /tmp --volumes-from=$cacheVolume -e TEST_DOCKERD -e SKIP_INTEGRATION_TESTS=0 -e BUILDKIT_REGISTRY_MIRROR_DIR=$VOL/registry --privileged $IID go test -v -run=".*/.*/worker=${{matrix.worker}}$" ${{ matrix.flags }} ${{ matrix.pkg }})
          ...
```
 We only need to specify the include for the exceptions where we want to add custom flags.  This drops the `./...` non-integration tests, it seems to me those should be run separately outside of the "integration test" matrix.

Reply:
Thanks for your input I will make some tests around this.